src/plugins/lua/luaplugin.h

```diff
@@ -79,9 +70,72 @@ class LUASHARED_EXPORT LuaPlugin : public Tiled::WritableMapFormat
     void writePolygon(LuaTableWriter &, const Tiled::MapObject *);
     void writeTextProperties(LuaTableWriter &, const Tiled::MapObject *);
 
-    QString mError;
-    QDir mMapDir;     // The directory in which the map is being saved
+    void mapDir(const QString& dir);
+
+private:
+    QDir mMapDir;
     Tiled::GidMapper mGidMapper;
+
+};
+
+class LUASHARED_EXPORT LuaPlugin : public Tiled::Plugin
+{
+    Q_OBJECT
+    Q_INTERFACES(Tiled::Plugin)
+    Q_PLUGIN_METADATA(IID "org.mapeditor.Plugin" FILE "plugin.json")
+
+public:
+    void initialize() override;
+};
+
+
+class LUASHARED_EXPORT LuaMapFormat : public Tiled::WritableMapFormat
+{
+    Q_OBJECT
+
+public:
+    explicit LuaMapFormat(QObject *parent = nullptr)
+        : WritableMapFormat(parent)
+    {}
+
+    bool supportsFile(const QString &fileName) const override;
+
+    bool write(const Tiled::Map *map, const QString &fileName) override;
+
+    QString nameFilter() const override;
+    QString shortName() const override;
+    QString errorString() const override;
+
+protected:
+    LuaUtilWriter mLuaWriter;
```

Comment:
I suggested to allocate the `LuaUtilWriter` on the stack. Was there a reason it should be a member instead?

Reply:
It has ``mMapDir`` and is used across calls e.g. ``writeProperties`` which seems to use the member assuming it has been set previously. I could be wrong, please feel free to correct me.

Reply:
I've pushed a change which puts it on the stack, since I wanted to make sure anyway that it would work. Now the `mMapDir` thing works even better I think, since by requiring it in the constructor we make sure that it is set appropriately. Only the static methods could be called without initializing it (though they aren't).

Reply:
Looks great :)